### PR TITLE
feat: rebuild SQLite when migrations occur

### PR DESCRIPTION
**Pull request by EvanHahn**
See #436.

### CODE REVIEW

File: multi-core-indexer-1.0.0-alpha.10.tgz
Comment (EvanHahn) [inline multi-core-indexer-1.0.0-alpha.10.tgz:1]: I plan to remove this once https://github.com/digidem/multi-core-indexer/pull/45 is merged.

File: src/mapeo-project.js [hunk 1]
```diff
@@ -139,11 +144,37 @@ export class MapeoProject extends TypedEmitter {
...
     this.#sqlite = new Database(dbPath)
     const db = drizzle(this.#sqlite)
-    migrate(db, { migrationsFolder: projectMigrationsFolder })
+    const migrationsBefore = tableCountIfExists(db, DRIZZLE_MIGRATIONS_TABLE)
+    migrate(db, {
+      migrationsFolder: projectMigrationsFolder,
+      migrationsTable: DRIZZLE_MIGRATIONS_TABLE,
+    })
+    const migrationsAfter = tableCountIfExists(db, DRIZZLE_MIGRATIONS_TABLE)
+    const reindex = migrationsBefore > 0 && migrationsAfter !== migrationsBefore
```
Comment (gmaclennan) [inline src/mapeo-project.js:156]: Maybe wrap this in our own migrate function for clarity? e.g. 
```
const didMigrate = migrate(db, { migrationsFolder: projectMigrationsFolder })
```

**Review by gmaclennan — CHANGES_REQUESTED** [review]
Ok looks good apart from one code clarity suggestion. However I think we should also include in this re-indexing of the `projectSettingsTable` (in `MapeoManager`) which I think needs the same treatment?

File: src/lib/drizzle-helpers.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Get the number of rows in a table using `SELECT COUNT(*)`.
+ * Returns 0 if the table doesn't exist.
+ *
+ * @param {BetterSQLite3Database} db
+ * @param {string} tableName
+ * @returns {number}
+ */
+export const tableCountIfExists = (db, tableName) =>
```
Comment (gmaclennan) [inline src/lib/drizzle-helpers.js:28]:
```suggestion
export const countTableRows = (db, tableName) =>
```
Naming of this one is hard! Not sure if this is better. The original name suggested to me that it is counting tables ("tableCount"), and only counting if something exists ("IfExists"), where-as I think what this is doing is counting rows in a table whether it exists or not.

Comment (gmaclennan) [reply src/lib/drizzle-helpers.js:28]: `safeCountTableRows()`?

File: src/lib/drizzle-helpers.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  db.transaction((tx) => {
+    const existsQuery = sql`
+      SELECT EXISTS (
+        SELECT 1
+        FROM sqlite_master
```
Comment (gmaclennan) [inline src/lib/drizzle-helpers.js:33]: Maybe use `sqlite_schema`? https://sqlite.org/schematab.html

Comment (gmaclennan) [reply src/lib/drizzle-helpers.js:33]: When you do `PRAGMA table_list` you see `sqlite_schema` as the only table.

Comment (EvanHahn) [reply src/lib/drizzle-helpers.js]: `sqlite_master` is an "historical" alias for `sqlite_schema`. I should be using `_schema`. Fixed in 574dca554b057d4f822c4279f5c47c6e0fa208d2.

File: src/lib/drizzle-helpers.js (hunk 2 above)
Comment (EvanHahn) [reply src/lib/drizzle-helpers.js]: Renamed it in f93b4a4dfbc45021e937003583c5d96c6c7edc0e. I also stopped exporting it in ae6e0c6d4d97b160f5336b79a578f80bc1aa7085.

File: src/mapeo-project.js (hunk 1 above)
Comment (EvanHahn) [reply src/mapeo-project.js]: Done in ae6e0c6d4d97b160f5336b79a578f80bc1aa7085.

**Comment by EvanHahn** [comment]
> Ok looks good apart from one code clarity suggestion. However I think we should also include in this re-indexing of the `projectSettingsTable` (in `MapeoManager`) which I think needs the same treatment?

Done in 5084bb2e7d853861ca4fb39807c4d9d21761fb72. (I didn't have an easy way to test this so I didn't.)

> Final thoughts, one way of doing this in the future would be to compare just the indexed tables in the `sqlite_schema` table, comparing the `sql` column, to detect changes. That way we would avoid re-indexing if we change the schema of a table that is not coming from our indexer.

That's a much better idea. I'll consider doing that in a followup.

**Review by gmaclennan — APPROVED** [review]

**Pull request merged**